### ISSUE TITLE
Race condition in taps.cljc?

### ISSUE THREAD

**Issue by rob-3** [jank-lang/clojure-test-suite#15]
I have been running the full test suite frequently while developing tests for `=`, and I nondeterministically get the following error:

```clojure
FAIL in (tapping) (taps.cljc:27)
expected: (= 104 (clojure.core/deref counter2))
  actual: (not (= 104 106))
```
I can reliably reproduce this by running:
```bash
while true; do
  lein test clojure.core-test.taps || break
done
```
and letting it run for a minute or two.

I have an M1 Pro and am running on ARM, so if you can't reproduce on x86 this may be why.

I've never used `tap>` before, but I see a `(sleep 50)` in the test that I imagine is the root cause.

**Comment by jeaye** [comment]
Great find. I know that @quoll struggled to get those tests working. Sounds like there are still some issues with it.

If you or @quoll are interested in digging into what's going on, let's do that. Otherwise, we can comment out the failing test with a link to this ticket and a brave hero will pick it up at some point.

**Comment by jeaye** [comment]
These have been disabled for now, to improve sanity while we grow our test coverage.

**Comment by jaidetree** [comment]
Looks like it's only the clojure version of the test that's flaky? I recall fixing the cljs version.

**Comment by jaidetree** [comment]
Wait @jeaye are these still flaky for you? They haven't failed in CI for months since last updated.

![Image](https://github.com/user-attachments/assets/6fc9b03f-a030-4dca-b273-3e31495711de)

They have been working consistently since https://github.com/jank-lang/clojure-test-suite/commit/a1660ba70a713e73b74ef0353b2bb4167e5e531f, I think this is just an old ticket?

**Comment by jeaye** [comment]
I haven't seen failures, but I wasn't convinced they were all resolved. Race conditions sometimes never show up in CI, since some race conditions only show up on faster machines (and CI machines are *slow*). So we can't treat CI as the only source of truth for this being fixed. 

I can enable them again, if you're confident that they're fixed. I'll just put a comment at the top of the file to point anyone who sees a failure here so we can know if it still happens.

**Comment by jaidetree** [comment]
> I haven't seen failures, but I wasn't convinced they were all resolved. Race conditions sometimes never show up in CI, since some race conditions only show up on faster machines (and CI machines are _slow_). So we can't treat CI as the only source of truth for this being fixed.

That's true but before they were fixed, they were flaky in CI. Now they have been reliable.

Anyway, ran the tests ~50 times each on my local MBP M3 Max, which is decently fast, and not once have the taps test failed. 

https://github.com/jank-lang/clojure-test-suite/commit/a1660ba70a713e73b74ef0353b2bb4167e5e531f was written in January 25th where as this issue was filed December 1st. It seems I forgot to update this issue.

**Comment by jaidetree** [comment]
Also look at the code, they are not going to be flaky. Instead of relying on a race condition that depends on computer speed like the previous version, the refactored version sets up an atom, and defines an `await-tap` function that pushes a promise that is only delivered when the tap queue gets to that item to deliver a payload to it.

**Comment by jeaye** [comment]
Nice. Thanks for following up on this. I'll enable those and we can close the ticket. 🙂
